evm_arithmetization/src/cpu/kernel/asm/mpt/storage/storage_write.asm

```diff
@@ -9,11 +9,26 @@ global sys_sstore:
     %stack (kexit_info, slot, value) -> (slot, kexit_info, slot, value)
     %sload_current
     %address
-    %stack (addr, current_value, kexit_info, slot, value) -> (addr, slot, current_value, current_value, kexit_info, slot, value)
+    %stack (addr, current_value, kexit_info, slot, value) -> (addr, slot, current_value, kexit_info, slot, value)
     %insert_accessed_storage_keys
-    // stack: cold_access, original_value, current_value, kexit_info, slot, value
-    %mul_const(@GAS_COLDSLOAD)
+    // stack: cold_access, value_ptr, current_value, kexit_info, slot, value
+    %jumpi(sstore_cold_access)
+    // stack: value_ptr, current_value, kexit_info, slot, value
+    MLOAD_GENERAL
+    // stack: original_value, current_value, kexit_info, slot, value
+    PUSH 0
+    // stack: gas, original_value, current_value, kexit_info, slot, value
+    %jump(sstore_after_cold_access)
+
+sstore_cold_access:
+    // stack: value_ptr, current_value, kexit_info, slot, value
+    DUP2 MSTORE_GENERAL
+    // stack: current_value, kexit_info, slot, value
+    DUP1
+    PUSH @GAS_COLDSLOAD
+    // stack: gas, original_value, current_value, kexit_info, slot, value
 
+sstore_after_cold_access:
```

Comment:
The name is a bit strange to me since you go there whether or not that was a cold access.

Reply:
Changed to `sstore_after_cold_access_check`